tests/test_lime_survey.py

```diff
@@ -233,11 +233,13 @@ def test_get_response_multiple_choice(self):
                 False,
                 False,
                 False,
-                False,
+                np.nan,
             ],
         ]
         response = self.survey.get_responses(self.multiple_choice_column, labels=False)
-        np.testing.assert_array_equal(expected_response, response.values[:2])
+        np.testing.assert_allclose(
+            expected_response, response.values[:2], equal_nan=True
```

Comment:
This error took me a long time to pin down...

As a result of having a `np.nan` now instead of purely boolean values, `response.values[:2]` returns an ndarray of dtype object, and the hidden `np.isnan()`  in `assert_array_equal` and `assert_allclose`  cannot handle object dtype, when `equal_nan` is set to True, hence throwing the error of unsafe coercion. Hence, `response.values[:2]` should be recast with `.astype(np.float64)` first before comparing with  `expected_response`

The same also goes for `assert_array_equal` on Line 252.

Is there a good reason to switch from `assert_array_equal` to `assert_allclose`? As far as I can see we are not comparing small numbers here, which is what `assert_allclose` is usually used for or?

Next time please make your best effort to ensure all the tests pass, because you are expected to be the one that knows the best the changes you have just made and can most easily diagnose the errors.

Reply:
Hi Dawei,

Thank you very much for reviewing my code, my apologies for the late replies, the past week has been hectic for me. You are right that `assert_array_equal` gave me problem with the `np.nan`, that is why I went with `assert_allclose` instead. I tried `np.testing.assert_array_equal` like you suggested, but it still raised an error saying that the `np.nan` are mismatched... What I did is first converting `expected_response` to an `np.array` with `dtype=object` and then comparing this with `response.values`, if I do not convert the expected_response from `list` to `array`, the test treats the boolean values as 1 and 0, and returns that the arrays are not equal...

Could you show me the inputs that you give to `np.testing.assert_array_equal` that made it work?

Thanks again for checking!

Reply:
Converting to object dtype wouldn't work, because two instances of `np.nan` as objects are different. In fact, an object-type array isn't even equal to itself (see screenshot below).
![image](https://user-images.githubusercontent.com/77344869/152694922-e3890e4a-987f-4559-8f1a-dd436133c52b.png)
Or if you try to make use of `equal_nan` in `np.testing.assert_allclose`, it complains about object dtype being not supported, as you already know.

Therefore, the only around this is that one should instead have both arrays in `np.float64` dtype and then compare. In fact, as soon as you recast the `responses.values` into float64, it doesn't matter if your `expected_response` is a list or an ndarray as both work with either `np.testing.assert_allclose` or `np.testing.assert_equal` without it complaining about `np.nan` (see screeshots below). 
![image](https://user-images.githubusercontent.com/77344869/152695128-47392df8-b1ab-423d-9be5-7a3488820013.png)

Of course, a list has no dtype attribute, but you see it works.
![image](https://user-images.githubusercontent.com/77344869/152695182-d4027a51-6139-4e3f-ba01-9102656bb684.png)

Also btw, boolean True and False are equivalent to 1 and 0 in python, so there is no need to worry about that (source: https://docs.python.org/3/reference/datamodel.html#index-10).

Give this a try, and if it still doesn't work, we can have a look together at the meeting tomorrow morning.

Reply:
Thanks, Dawei, for the really descriptive explanation. I have modified the line 240 to use `np.testing.assert_equal` as you suggested. Also, I had to change line 252/253 to recast the `response.values` to be type `np.float64`. But somehow in this case `np.testing.assert_array_equal` does not give any problem while comparing arrays with `np.nan`, so I'm really confused by this function. I have committed my changes.